### PR TITLE
dev-faucet

### PR DESCRIPTION
**Pull request by garbageslam**
This adds a thin HTTP server called `mobilecoind-dev-faucet`, similar to `mobilecoind-json` but much smaller, which functions as a faucet and can be deployed in dev networks.

This an alternative way that clients can get funds to run tests, which may simplify a lot of the things we do in deployment around funding a large number of accounts of several varieties as part of the deployment. 

---

The server has a background thread which periodically checks if it should split off smaller TxOut's from the main faucet balance, and queues pre-split TxOut's in a tokio queue.

It should be able to handle a relatively high volume of concurrent faucet requests, so if e.g. SDK unit tests hit the faucet many times it won't fail.

I tested this version locally using the local_network.py script, it seems to be working as expected.

### CODE REVIEW

File: mobilecoind-dev-faucet/README.md
Comment (eranrund) [inline mobilecoind-dev-faucet/README.md:61]:
```suggestion
$ curl -s localhost:9090/ -d '{"b58_address": "c7f04fcd40d093ca6578b13d790df0790c96e94a77815e5052993af1b9d12923", "token_id": "1"}' -X POST -H 'Content-type: application/json'
```

File: mobilecoind-dev-faucet/src/bin/main.rs
Comment (eranrund) [inline mobilecoind-dev-faucet/src/bin/main.rs:152]: I don't think this is the right call...

**Review by eranrund — COMMENTED** [review]
Looks good so far, thank you for doing this!

File: mobilecoind-dev-faucet/src/bin/main.rs
Comment (garbageslam) [reply mobilecoind-dev-faucet/src/bin/main.rs:152]: i'm trying to base it on the other PR, so in the `last-block-info` PR as it currently exists, I think this would be the right call. but if we redo that then we might have to change this

File: mobilecoind-dev-faucet/README.md
Comment (garbageslam) [reply mobilecoind-dev-faucet/README.md:61]: thank you!

File: mobilecoind-dev-faucet/src/bin/main.rs
Comment (garbageslam) [reply mobilecoind-dev-faucet/src/bin/main.rs:152]: oh i'm sorry, this is the wrong call, you're right, I'm supposed to call the get tx status thing. whoops

File: mobilecoind-dev-faucet/src/worker.rs
Comment (eranrund) [inline mobilecoind-dev-faucet/src/worker.rs:497]: Might need to account for the fee as well here for this to be accurate.

File: mobilecoind-dev-faucet/src/worker.rs
Comment (eranrund) [inline mobilecoind-dev-faucet/src/worker.rs:520]: `.into()` should work instead of `RepeatedField::from_vec`

**Review by eranrund — COMMENTED** [review]
Some small nits but other than that looking good.

File: mobilecoind-dev-faucet/src/data_types.rs
Comment (garbageslam) [reply mobilecoind-dev-faucet/src/data_types.rs:47]: good idea

File: mobilecoind-dev-faucet/src/bin/main.rs
Comment (iamalwaysuncomfortable) [inline mobilecoind-dev-faucet/src/bin/main.rs:79]: Maybe more clarification these are the payout amounts?

```suggestion
    pub faucet_payout_amounts: HashMap<TokenId, u64>,
```

OR 

```suggestion
    pub faucet_payouts: HashMap<TokenId, u64>,
```

File: mobilecoind-dev-faucet/src/worker.rs
Comment (iamalwaysuncomfortable) [inline mobilecoind-dev-faucet/src/worker.rs:19]: It kind of seems like we're doing a custom implementation of futures and an executor here? This is pretty impressive what you made here - but a lot of it also seems like it could be done using Rust's normal async primitives + an executor like Tokio/Block-on/Async-std.

We're avoiding using the Tokio executor - but why exactly? It seems like there's a boilerplate could be avoided by functionality already existing in Tokio and other executors like spawned tasks, threads, and a ton of helper structs that provide things like queues that can work in an async context:
https://docs.rs/futures/0.3.21/futures/future/index.html
https://docs.rs/futures/0.3.21/futures/stream/index.html

I.e. like these for qeueing up tasks:
https://docs.rs/futures/0.3.21/futures/stream/struct.FuturesUnordered.html
or 
https://docs.rs/futures/0.3.21/futures/stream/struct.FuturesOrdered.html

I think we lose some level of maintaibility here, ability to work with native rust concurrency, and also some pre-done optimization that other executors have. 

If this already works and the faucet is ready to go, we can/should go with this, but it could in the future potentially we could go towards a more "native async" approach.

Comment (garbageslam) [reply mobilecoind-dev-faucet/src/worker.rs:19]: I don't think I made a custom implementation of futures or an executor here -- I think all I did was create a background worker thread. 

I considered doing it another way:
* Replace `std::thread::spawn` with `tokio::task::spawn`
* Replace `std::thread::sleep` with `tokio::task::sleep(..).await`.

The main reason that I didn't do that is because I wasn't sure if I'm allowed to await futures that come from grpcio in a task spawned by the tokio stuff (that was partly why I asked questions earlier).

Anyways, it seems like there are fewer questions that arise if I just make it an `std::thread`, and the worker doesn't really need parallelism or more than one thread. So I wasn't sure what was gained by making it a tokio task instead.

Note also that, most of our rust web services right now, including consensus, fog ingest, fog view, etc. all spawn background threads this way and communicate either using queues, or sometimes just std lib mutexes. So I guess I just felt very comfortable making the background worker an std thread.

The other thing was, by making it a tokio task, it's possible that the worker thread can be starved if there are a lot requests that come in, because all the OS threads could be assigned to the rocket task I guess. Maybe it doesn't really matter. But it seemed a little easier to think about in what circumstances the worker could get blocked if I don't make it use the same thread pool as the HTTP server.

---

I'm sure that there are things that can be improved about the worker thread if it is async, the main one being, instead of sleeping a fixed amount of time, you could do some kind of select over the one-shot receivers and the timer, so that it wakes up immediately if there might be something to do. You could also do that using an std condition variable I think but select is better for various reasons. But I'm not sure if having the worker thread wake up immediately in those scenarios matters. And just using `std::thread` this way seemed simpler to me.

---

Anyways if you see a major benefit to changing it to `tokio::Task` then I'm happy to change this, but I'm curious if in your mind this is more than substituting `std::thread::spawn` with `tokio::task` and `std::thread::sleep` with `tokio::time::sleep.await`.

I don't see that the Streams stuff is relevant here -- if it became necessary to use those APIs, we could turn the mpsc queue endpoint into a stream and then use it that way: https://docs.rs/tokio-stream/0.1.8/tokio_stream/wrappers/struct.ReceiverStream.html

Comment (iamalwaysuncomfortable) [reply mobilecoind-dev-faucet/src/worker.rs:19]: Also thank you for doing this, this is an *excellent* feature

File: mobilecoind-dev-faucet/src/data_types.rs
Comment (remoun) [inline mobilecoind-dev-faucet/src/data_types.rs:90]: Nit: Sort
```suggestion
#[derive(Debug, Default, Deserialize, Serialize)]
```

File: mobilecoind-dev-faucet/src/bin/main.rs
Comment (remoun) [inline mobilecoind-dev-faucet/src/bin/main.rs:67]: Consider moving this `State` into a `mod` so we could test it

File: mobilecoind-dev-faucet/src/bin/main.rs
Comment (remoun) [inline mobilecoind-dev-faucet/src/bin/main.rs:351]: Consider extracting a helper for this Rocket/server setup, a la https://github.com/mobilecoinfoundation/mobilecoin/blob/master/fog/overseer/server/src/server.rs

File: mobilecoind-dev-faucet/src/worker.rs
Comment (remoun) [inline mobilecoind-dev-faucet/src/worker.rs:218]: Please move this thread worker logic into a separate method

File: mobilecoind-dev-faucet/src/worker.rs
Comment (remoun) [inline mobilecoind-dev-faucet/src/worker.rs:453]: Please document what this check is for.

**Review by remoun — COMMENTED** [review]
LGTM -- one question, otherwise nits.

File: mobilecoind-json/Cargo.toml
Comment (remoun) [reply mobilecoind-json/Cargo.toml:32]: Right, so `mc-util-serial` becomes a `[dependency]`, which you already added, which makes this `[dev-dependency]` redundant :)

File: mobilecoind-dev-faucet/src/bin/main.rs
Comment (garbageslam) [reply mobilecoind-dev-faucet/src/bin/main.rs:71]: After looking at this, I guess I'm not sure what the point of this is.

If I want to test the assembly of the server and rocket, then I think one way is, in `mobilecoind-dev-faucet/tests`, I could add integration tests that start the server by running the compiled binary, and then poke it over http etc. But this will also be quite complex because you need a consensus network for this test to work. So at some point it seems easier to make this like one of the tests that runs `local_network.py`.

I guess I don't see that it's interesting to build another binary that starts another http server using rocket, with the same routes as this one, but isn't actually the same binary. So if I'm going to make a re-usable helper out of this, who is the other user that is consuming it?

I think for the fog overseer there is a difference that, the integration tests for that server do actually start a bunch of fog ingest server on the stack in the test code: 

https://github.com/mobilecoinfoundation/mobilecoin/blob/9d7f88efe181ee498c07f62767ff56aed18a032c/fog/overseer/server/tests/inactive_outstanding_key_idle_does_not_have_key.rs#L30
https://github.com/mobilecoinfoundation/mobilecoin/blob/9d7f88efe181ee498c07f62767ff56aed18a032c/fog/overseer/server/tests/utils/mod.rs#L27

But I don't think this is currently how we test the consensus network + mobilecoind. Consensus network has a ton of threads and starting 5 consensus nodes on the stack isn't really going to work. We used to have a lot of tests like that and it was pretty flaky and eventually got removed. (Maybe, we should test fog ingest clusters differently too, but our approach there seems to be working, so if it ain't broke maybe don't fix it?)

Comment (remoun) [reply mobilecoind-dev-faucet/src/bin/main.rs:71]: Fair enough. I thought it might be interesting to exercise the faucet with a mock server, but I see your point that this is more akin to `local_network.py`

Comment (garbageslam) [reply mobilecoind-dev-faucet/src/bin/main.rs:71]: I see -- I think we could make a mock of mobilecoind in a tests/ file, and test the faucet using this:

https://doc.rust-lang.org/cargo/commands/cargo-test.html#target-selection

> Binary targets are automatically built if there is an integration test or benchmark. This allows an integration test to execute the binary to exercise and test its behavior. The CARGO_BIN_EXE_<name> [environment variable](https://doc.rust-lang.org/cargo/reference/environment-variables.html#environment-variables-cargo-sets-for-crates) is set when the integration test is built so that it can use the [env macro](https://doc.rust-lang.org/std/macro.env.html) to locate the executable.

I think it has some benefit, like it's actually testing the final binary, and it doesn't have to build the rocket part again for the test

File: mobilecoind-dev-faucet/src/worker.rs
Comment (iamalwaysuncomfortable) [reply mobilecoind-dev-faucet/src/worker.rs:19]: Yeah I'm good with the way this is. It works (and is impressive engineering..) and it solves a pretty significant problem in our release cycles, so it should go in.

As to the engineering philosophy bit, I think there's a bit of an art to mixing threads and executors. Std::thread + an async executor is pretty common in practice I think and a valid approach. Sync code can pass as effectively async through the mspc channels between threads because the channels act as a stream that that queue up done tasks as is done here.  But there are other ways and I think my main point is let's huddle more as a team on async design in the design phase of things like this as I think that's a path towards having a team mastery of the async paradigm.

Thanks for building this!

File: mobilecoind-dev-faucet/src/bin/main.rs
Comment (remoun) [inline mobilecoind-dev-faucet/src/bin/main.rs:28]: Optional/Equivalent: Consider moving this error handling into an `impl From<Result<mc_mobilecoind_api::SubmitTxResponse, String> for JsonSubmitTxResponse`, then this method can be
```suggestion
    Json(state.handle_post(&req).await.into())
```

File: mobilecoind-dev-faucet/src/lib.rs
Comment (garbageslam) [reply mobilecoind-dev-faucet/src/lib.rs:285]: i tried this in an earlier revision, but `convert_balance_pair` only either works with references, or not with references. i figured having less code is more valuable than avoiding some copies, since we need it to work somewhere else with references.

File: mobilecoind-dev-faucet/src/lib.rs
Comment (garbageslam) [reply mobilecoind-dev-faucet/src/lib.rs:241]: hmm, so this error occurs if the one-shot receiver was dropped by the worker thread before we sent this.

i feel like, this represents some kind of internal error or race, but we still submitted a transaction successfully. if we propagate this error, then we are telling the faucet user that the tx failed, but that may not be true -- at this point I think we submitted something successfully but had an internal error passing a message.

i feel like the right thing to do is log but not propagate, because i think the right thing for the faucet to do is return `success: true` and let the client track what happens the payment if they want to

File: mobilecoind-dev-faucet/src/worker.rs
Comment (garbageslam) [reply mobilecoind-dev-faucet/src/worker.rs:262]: I guess that it will log errors at a rate of `worker_poll_period`, unless / until mobilecoind recovers

That seems like desirable behavior, what would you rather see?

Comment (garbageslam) [reply mobilecoind-dev-faucet/src/worker.rs:262]: We could backoff on the timeout so that we don't log at faster than 1 Hz or something?

File: mobilecoind-dev-faucet/src/worker.rs
Comment (garbageslam) [reply mobilecoind-dev-faucet/src/worker.rs:502]: I guess I'm not sure if it's interesting to try to handle this error

This error occurs if we try to add a utxo to the queue, but we cannot. The queue is an unbounded queue, so according to docs it only happens if `close` is called or if the receiver end is dropped.

Close is not called, so the only way this can happen is if the Worker handle struct is dropped. But drop for that function calls join for the handle to this thread. So this really should never happen.

If the receiver end of the queue is dropped, the worker thread cannot do anything useful anymore. If we make this error be map err and returned to caller, then the error will just be logged infinitely in a loop. It might be reasonable to try to say, if this error is encountered, we should try to exit this thread gracefully. But this will also be a lot of plumbing, and I'm not sure it really improves the overall health of the program, because the faucet will stop working if the worker has joined, and the faucet doesn't currently have any way to restart the worker.

Actually I guess the most sensible thing to do is panic, since the worker thread expects this not to ever fail, and the worker thread can't really do anything useful anymore if the queue is closed. I would be open to changing this to a panic

Comment (remoun) [reply mobilecoind-dev-faucet/src/worker.rs]: Panicking for this exceptional case SGTM

File: mobilecoind-dev-faucet/src/data_types.rs
Comment (eranrund) [inline mobilecoind-dev-faucet/src/data_types.rs:17]: I think you can skip the `Option<>` and use `#[serde(default)]`

File: mobilecoind-dev-faucet/src/lib.rs
Comment (eranrund) [inline mobilecoind-dev-faucet/src/lib.rs:258]: I might be wrong, but if I understand correctly whats happening here then it might be worth noting that the request handler is going to be executed from a tokio runner context, while this await is going to block on whatever runner grpcio is using. As such maybe its not worth using the `async` variant of the API client?

**Review by eranrund — APPROVED** [review]
LGTM, although there might be an issue with mixing the grpcio calls and tokio executor. The recommended way according to some issues I found on the grpcio GH repository is to use some form of channel to bridge between the two.
With that said, we don't particularly care about slowing down the http request handling here so I think it is okay if they block for the duration of the mobilecoind GRPC calls.
Not blocking this on approval since this isn't critical code, but I do suggest documenting this and maybe not using the `_async` call variants at all inside tokio async functions.

File: mobilecoind-dev-faucet/src/lib.rs
Comment (remoun) [reply mobilecoind-dev-faucet/src/lib.rs:258]: No, each suspended Future tracks a waker to resume when it should be polled again, e.g. when data arrives (impls have to call the waker when `Future::poll` returns `Poll::Pending`). They can be running on different executors, including two Tokio runtimes. So the gRPC futures will likely be driven by the gRPC executor (in C) and then these handlers would use that data while running in a Tokio executor

Comment (garbageslam) [reply mobilecoind-dev-faucet/src/lib.rs:258]: I had similar thoughts and I asked in the fog-core team standup (since there are a lot of async experts on that team now) if this is a bug, to my understanding, this isn't a bug and this is the way we're supposed to do it.

Here's busyjay on the issue: https://github.com/tikv/grpc-rs/issues/520

* tokio and grpcio have different thread pools and different executors in this program, the grpcio one is based in some C code.
* calls to `tokio::task::spawn` (or any tokio function that calls that) must be made in the context of the tokio executor, tokio will panic like `grpc-poll-0' panicked at 'not currently running on the Tokio runtime.` (https://github.com/tikv/grpc-rs/issues/338)
* awaiting a grpcio future in a tokio context is not bad though, if a tokio task goes to sleep on a grpcio future, i think the grpcio executor will ultimately call the "waker" from tokio via a generic and compatible API, and this is expected to work
* using an async API here is good because in theory this thread can be used by tokio to handle more requests
* even if we switched to calling a sync API, it is likely that the sync API works by calling the async api and then using `.await` (https://docs.rs/grpcio/latest/src/grpcio/client.rs.html#41)

that last point may be a good argument for switching to calling the sync version again, since it is calling await under the hood for us anyways, but we would have less code / less error handling boilerplate and still reap the benefits of async

Comment (remoun) [reply mobilecoind-dev-faucet/src/lib.rs:258]: > * using an async API here is good because in theory this thread can be used by tokio to handle more requests

That's precisely the benefit of async APIs vs blocking.

> * even if we switched to calling a sync API, it is likely that the sync API works by calling the async api and then using `.await` (https://docs.rs/grpcio/latest/src/grpcio/client.rs.html#41)
> 
> that last point may be a good argument for switching to calling the sync version again, since it is calling await under the hood for us anyways, but we would have less code / less error handling boilerplate and still reap the benefits of async

Note that they're calling `block_on` to block the current thread, which is very different from `.await`

**Review by remoun — APPROVED** [review]

File: mobilecoind-dev-faucet/src/lib.rs
Comment (garbageslam) [reply mobilecoind-dev-faucet/src/lib.rs:136]: okay sounds good

**Pull request merged**
